canary/MeasureTransferRate.cpp

```diff
@@ -0,0 +1,484 @@
+                        AWS_LS_CRT_CPP_CANARY,
+                        "Transfer finished with error %d: '%s'",
+                        errorCode,
+                        aws_error_debug_str(errorCode));
+                }
+
+                --numInProgress;
+                ++numCompleted;
+
+                transferCompletedSignal.notify_one();
+            };
+
+        AWS_LOGF_INFO(
+            AWS_LS_CRT_CPP_CANARY,
+            "Beginning transfer %d - Num Concurrent:%d/%d  Total:%d/%d",
+            i,
+            numInProgress.load(),
+            numConcurrentTransfers,
+            numCompleted.load(),
+            numTransfers);
+
+        transferFunction(i, std::move(key), transport, std::move(notifyTransferFinished));
+
+        // Wait if number of inprogress transfers is not less than number of concurrent transfers.
+        std::unique_lock<std::mutex> guard(transferCompletedMutex);
+        transferCompletedSignal.wait(guard, [&numInProgress, numConcurrentTransfers]() {
+            return numInProgress.load() < numConcurrentTransfers;
+        });
+    }
+
+    // Wait until all transfers have completed.
+    std::unique_lock<std::mutex> guard(transferCompletedMutex);
+    transferCompletedSignal.wait(
+        guard, [&numCompleted, numTransfers]() { return numCompleted.load() >= numTransfers; });
+
+    // If this is true, then we are in fork mode, and are executing in a child process,
+    // so report finished to the parent process.
+    if (m_canaryApp.GetOptions().isChildProcess)
+    {
+        m_canaryApp.WriteToParentProcess(finishedKey.c_str(), "done");
+    }
+
+    transport->PurgeConnectionManagers();
+}
+
+void MeasureTransferRate::MeasureHttpTransfer()
+{
+    String endpoint = m_canaryApp.GetOptions().httpTestEndpoint.c_str();
+
+    Aws::Crt::Http::HttpHeader hostHeader;
+    hostHeader.name = ByteCursorFromCString("host");
+    hostHeader.value = ByteCursorFromCString(endpoint.c_str());
+
+    Http::HttpClientConnectionManagerOptions connectionManagerOptions;
+    connectionManagerOptions.ConnectionOptions.HostName = endpoint;
+    connectionManagerOptions.ConnectionOptions.Port = m_canaryApp.GetOptions().sendEncrypted ? 443 : 5001;
+    connectionManagerOptions.ConnectionOptions.SocketOptions.SetConnectTimeoutMs(3000);
+    connectionManagerOptions.ConnectionOptions.SocketOptions.SetSocketType(AWS_SOCKET_STREAM);
+    connectionManagerOptions.ConnectionOptions.InitialWindowSize = SIZE_MAX;
+
+    if (m_canaryApp.GetOptions().sendEncrypted)
+    {
+        aws_byte_cursor serverName = ByteCursorFromCString(endpoint.c_str());
+        auto connOptions = m_canaryApp.GetTlsContext().NewConnectionOptions();
+        connOptions.SetServerName(serverName);
+        connectionManagerOptions.ConnectionOptions.TlsOptions = connOptions;
+    }
+
+    connectionManagerOptions.ConnectionOptions.Bootstrap = &m_canaryApp.GetBootstrap();
+    connectionManagerOptions.MaxConnections = 5000;
```

Comment:
Config param?

Reply:
Probably should actually be based off of a handful of other constants.  I'll try to figure out a better way to define this.